### ISSUE TITLE
Feature Request / How to: Pipe Verbose Logs to file

### ISSUE THREAD

**Issue by luke-hill** [taf2/curb#378]
As the title suggests, how can we do this??

I found an article here: https://stackoverflow.com/questions/5427454/how-do-i-pipe-or-redirect-the-output-of-curl-v About doing it when running the curl client directly on the command line.

But it would be really good when creating a new easy handle (Which will have the `verbose?` method return `true` or `false` be able to also have a `output_file` - By default this would still remain as `$stdout`

Any thoughts / pointers if this already exists?

@robuye - I know you've been active on here recently.

**Comment by robuye** [comment]
Verbose output from `curb` goes to STDERR so you could capture it to a file this way:

```ruby
require 'curb'

$stderr.reopen("/tmp/err.txt", "w")
Curl::Easy.perform("http://www.google.co.uk") do |curl|
  curl.verbose = true
end
```

I had to use `reopen` or it wouldn't work otherwise.  `$stderr = StringIO.new` just doesn't work with curb and I didn't want to dig into libcurl.

Better you could use the `on_debug` callback, it should be much more reliable to capture verbose output than hacking `$stderr`. There are other callback that may work for you, so have a look, maybe `on_header` would suffice.

```

**Comment by luke-hill** [comment]
Would re-directing stderr affect other services such as jenkins console logging?

Wouldn't it be desirable to have this as a config option inside the `curb` library? Feels so to me with every other "logging" utility out there able to pipe to a file.

I'm happy to learn/pair on something if feasible.

**Comment by robuye** [comment]
> Would re-directing stderr affect other services such as jenkins console logging?

STDERR is global so yes, maybe, but it's rather unlikely if you restore original stderr after the transfer. If you use multiple threads / processes / callbacks then do not do it. I wouldn't recommend it either way, but I mention it because it's technically possible.

> Wouldn't it be desirable to have this as a config option inside the `curb` library? Feels so to me with every other "logging" utility out there able to pipe to a file.

Any pipe redirection is being handled by the OS, the `$stdout` & `$stderr` are just a File Descriptors in general and can be changed as in the first example snippet. It's good to understand a concept of File Descriptors in linux because it applies universally to any process and FD can point to a remote host over network, to a file or to a device (ie terminal) and it's interchangeable. That makes any process capable of logging (technically writing bytes) to any file or device.

That said being able to control where the verbose output from curb goes without unsafe workarounds would be great.

> I'm happy to learn/pair on something if feasible.

Thanks! `libcurl` provides `CURLOPT_STDERR` to set arbitrary file handle to use as `stderr` and if it's not set it will use default stderr (from C, not `$stderr` global from Rubyland). This option accepts a raw C `* FILE` pointer and it's a bit problematic because it's not always available on IO-like objects (ie `StringIO` doesn't have it). I will open a P/R with sufficient implementation to work with files and if it still appeals you're welcome to build on top of that ie. to accept Strings or Tempfile.

**Comment by luke-hill** [comment]
Merged and released in `0.9.10`
